### PR TITLE
Move the WeBWorK::PG module to PG and improve PG's local configuration.

### PR DESCRIPTION
**Pull request by drgrice1**
Remove the ALWAYS_SHOW_HINT_PERMISSION_LEVEL and ALWAYS_SHOW_SOLUTION_PERMISSION_LEVEL settings.  Whether hints and solutions are shown is now determined by the front end.

The pg flag/variable $showHint has been removed.  The front end has complete control on when hints are shown.  Not the problem author.

Also the options to use knowls for solutions and hints have been removed.  Always use knowls for solutions and hints.

Showing the various pieces of debugging information for a problem is now controlled with translation options.  The front end is responsible for converting permissions into these debugging options.  The VIEW_PROBLEM_DEBUGGING_INFO special pg environment variable was removed and is one of these debugging options.

Many unnecessary PG environment variables have been removed.  One of these is the session key.

Remove reduced scoring entirely from PG.  That is the responsibility of the front end to implement.

Pg now has no dates and and no permissions.

The color.c file has been removed.  The method in that file has been reimplemented in Chromatic.pm using perl.  It is perhaps a bit slower, but not significantly in the scope of rendering a problem.  This can be tested with the problems
    Library/NAU/setGraphTheory/FindChromaticNumber.pg,
    Library/NAU/setGraphTheory/FindRoomColor.pg, and
    Library/NAU/setGraphTheory/RoomSchedule.pg
which happen to be all known problems that use the Chromatic module.

The image generator has been rewritten to remove the dvipng_align hack and to allow the usage of sqlite for a database backend for storing dvipng depths.

Note that PGalias no longer uses the course name, user login, set number, and problem number for generating resource UUID's.  It now only uses the problem seed, psvn, and problemUUID.  It is the caller's responsibility of ensuring that problemUUID guarantees the necessary uniqueness.

Some unit tests were added for this.

The t/macros/load_macros.t file tests that all macros load without errors or warnings. There is one macro that will not work in this test and it is skipped. That is the answerDiscussion.pl macro.  It requires a file that is not part of the PG distribution.  It also has invalid `defined(@array)` calls.

Warnings found in other macro files were fixed.

The PG_module_list.pl macro was deleted.  That is clearly obsolete and will not work.

There are also some tests added in t/pg_problems that test that rendering with the WeBWorK::PG module works and demonstrate how to use it in unit tests.

There was also a "memory bleed" issue that was found with the Applet.pm module and the AppletObjects.pl macro caused by extending the Applet package in the macro.  Instead of extending the Applet package, a derived package (PGApplet) is created and used in the macro.  The GeogebraWebApplet package needs to derive from the PGApplet package, and so it needs to be moved to the macro.  Also, the Applet.pm package no longer derives from the PGcore package.  It doesn't need to as it uses none of the PGcore module methods.  A similar issue occurs with the tableau.pl macro, but that was not fixed.  So see the issue here, add a `for my $i (0 .. 1)` loop around the code inside the subtest in t/macros/load_macros.t.  When the tableau.pl macro is loaded the second time, a warning will be issued.

### CODE REVIEW

**Comment by drgrice1** [comment]
Note that the added tests use the Test2::Suite (via the Test2::V0 bundle).  So pull request #702 is needed to regain uniformity for the tests.

**Comment by pstaabp** [comment]
I'm starting to dig into this.  First of all, it looks like we'll need to update the database tables in the admin course.

Overall, things look great.  I have some problem sets for testing over the past year or so and old Flash Applets give the following error: `Can't locate object method "insertAll" via package "Applet" at line 85 of (eval 5768)` instead of saying that they are no longer deprecated.

**Comment by drgrice1** [comment]
Yes, to test this you will need to upgrade database tables.

This flash applets error is fixed.  You will need to reset the branch (or delete it and check it out again) to get the changes.

**Comment by pstaabp** [comment]
> You will need to reset the branch (or delete it and check it out again) to get the changes.

Odd. Still seeing this and have tried deleting and clearly back on both of these branches.

**Comment by pstaabp** [comment]
On the chromatic problems, On the latter two you listed above, I'm getting the following errors:
```
Use of uninitialized value in numeric gt (>) at [PG]/lib/Chromatic.pm line 275
Use of uninitialized value in numeric eq (==) at [PG]/lib/Chromatic.pm line 275
```

**Comment by drgrice1** [comment]
Make sure you reset the pg pull request to fix the flash issue, not this one.

**Comment by drgrice1** [comment]
I am not seeing the chromatic errors.

**Comment by drgrice1** [comment]
What seed do you have for the chromatic problems?  Maybe it is seed specific.

**Comment by drgrice1** [comment]
Okay, I didn't actually commit the changes to AppletObjects.pl.  Now they are in.  Sorry.

I changed seeds, and now I am seeing the use of uninitialized values for the last chromatic problem.  I will look into it.

**Comment by pstaabp** [comment]
Applets is fixed now.  :)

**Comment by pstaabp** [comment]
Also, wanted to note, that in the `PG_ROOT` directory, `prove -r t` runs everything successfully.  There's a number of test problems in the `t` directory that should either be removed/moved or have tests that run them.  (not for this PR though.)

**Comment by drgrice1** [comment]
The chromatic errors are fixed.  It was a faulty initialization.

**Comment by drgrice1** [comment]
Yeah, the t directory needs to be cleaned up.  TODO

**Review by drdrew42 — COMMENTED** [review]
Mostly inconsequential feedback from initial read-through. Testing forthcoming.

File: lib/WeBWorK/PG/Environment.pm [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		$pg_envir->{directories}{permitted_read_dir} = $ce->{webwork_courses_dir};
+		$pg_envir->{directories}{equationCache}      = $ce->{webworkDirs}{equationCache};
+		$pg_envir->{externalPrograms}                = $ce->{externalPrograms};
+		$pg_envir->{URLs}{equationCache}             = $ce->{webworkURLs}{equationCache};
+		$pg_envir->{equationCacheDB}                 = $ce->{webworkFiles}{equationCacheDB};
+	}
+
+	# Note that placeholders used in $pg_envir->{URLs}{html}, $pg_envir->{directories}{OPL}, and
+	# $pg_envir->{directories}{Contrib} are set on the first iteration, and those carry over to anywhere those
+	# placehoders are are used in other settings on the second iteration.
```
[the remark this answers is not in the document]
Comment (drgrice1) [reply lib/WeBWorK/PG/Environment.pm:103]: Yeah.  The `directory/OPL` and `directory/Contrib` settings are the values for the `OPL_dir` and `Contrib_dir` placeholders.  If those settings have placeholders (they have the $render_root placeholder in the renderer's pg_config.yml file, for example), then those placeholders are replaced on the first pass, and then those matriculate on to the places that the OPL_dir and Contrib_dir placeholders are used elsewhere on the second pass.

**Comment by drgrice1** [comment]
It would be good to get this in soon so that we can progress with other pull requests that depend on it.

**Comment by drgrice1** [comment]
This now includes the improved warning handling that we discussed.

**Comment by drgrice1** [comment]
Uhh, did not mean to close with comment.

**Pull request merged**